### PR TITLE
[CARBONDATA-3222]Fix dataload failure after creation of preaggregate datamap on main table with long_string_columns

### PR DESCRIPTION
**Pull request by shardul-cr7**
This PR is to Fix dataload failure after creation of preaggregate datamap on main table with long_string_columns.

Dataload is gettling failed because child table properties are not getting modified according to the parent table for long_string_columns.
This occurs only when long_string_columns is not specified in dmproperties for preaggregate datamap but the datamap was getting created and data load was failing. This PR is to avoid the dataload failure in this scenario.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
        added a testcase
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2110/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10364/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2112/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10366/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2317/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2114/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10368/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2115/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala [hunk 1]
```diff
@@ -333,6 +333,36 @@ class VarcharDataTypesBasicTestCase extends QueryTest with BeforeAndAfterEach wi
...
+         |""".
+        stripMargin)
+
+    sql(
+      s"""
+         | LOAD DATA LOCAL INPATH '$inputFile' INTO TABLE $longStringTable
+         | OPTIONS('header'='false')
+       """.stripMargin)
+
+    sql(s"drop table if exists $longStringTable")
```
Comment (qiuchenjian) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala]: Better to add a assert for this test case

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 2]
```diff
@@ -126,6 +126,12 @@ case class PreAggregateTableHelper(
         newLongStringColumn.mkString(","))
     }
 
+    //Add long_string_columns properties in child table from the parent.
+    tableProperties
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: @shardul-cr7 Here you are copying all the long String column...in case of pre aggregate data map may be it will not have all the columns of maintable, its better to add long string properties only for those column which are present in case of pre aggregate datamap......
example:  maintable long string columns: column1, column2, column3
datamap1: column1....set long string property for column1
datamap2:count(column1) no need to set if any udf, udaf is present then no need to set this property

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2320/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10369/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2140/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2346/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10394/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2144/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10398/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala (hunk 1 above)
Comment (shardul-cr7) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala]: Added!

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala (hunk 2 above)
Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Done. @kumarvishal09 please review.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 3]
```diff
@@ -110,22 +109,42 @@ case class PreAggregateTableHelper(
...
-    val longStringColumn = tableProperties.get(CarbonCommonConstants.LONG_STRING_COLUMNS)
+    // If longStringColumn is not present in dmproperties then we take long_string_columns from
+    // the parent table.
+    var longStringColumn = tableProperties.get(CarbonCommonConstants.LONG_STRING_COLUMNS)
+    val longStringColumnInParents = parentTable.getTableInfo.getFactTable.getTableProperties.asScala
+      .getOrElse(CarbonCommonConstants.LONG_STRING_COLUMNS, "").split(",").map(_.trim)
+    var varcharDatamapFields = ""
+    fieldRelationMap foreach (fields => {
+      val aggFunc = fields._2.aggregateFunction
+      if (aggFunc == "") {
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Please use isEmpty

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 4]
```diff
@@ -110,22 +109,42 @@ case class PreAggregateTableHelper(
...
     if (longStringColumn != None) {
       val fieldNames = fields.map(_.column)
-      val newLongStringColumn = longStringColumn.get.split(",").map(_.trim).map{ colName =>
+      val newLongStringColumn = longStringColumn.get.split(",").map(_.trim).map { colName =>
         val newColName = parentTable.getTableName.toLowerCase() + "_" + colName
         if (!fieldNames.contains(newColName)) {
           throw new MalformedDataMapCommandException(
             CarbonCommonConstants.LONG_STRING_COLUMNS.toUpperCase() + ":" + colName
-              + " does not in datamap")
+            + " does not in datamap")
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: revery this change

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 5]
```diff
@@ -110,22 +109,42 @@ case class PreAggregateTableHelper(
...
+        relationList.foreach(rel => {
+          rel.foreach(col => {
+            if (longStringColumnInParents.contains(col.parentColumnName)) {
+              varcharDatamapFields += col.parentColumnName + ","
+            }
+          })
+        })
+      }
+    })
+    if (varcharDatamapFields.size != 0) {
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: remove this code, covert comma separated string from seq

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 6]
```diff
@@ -110,22 +109,42 @@ case class PreAggregateTableHelper(
...
             CarbonCommonConstants.LONG_STRING_COLUMNS.toUpperCase() + ":" + colName
-              + " does not in datamap")
+            + " does not in datamap")
         }
         newColName
       }
       tableProperties.put(CarbonCommonConstants.LONG_STRING_COLUMNS,
         newLongStringColumn.mkString(","))
     }
-
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: revert this change

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala (hunk 5 above)
Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Done!

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 7]
```diff
@@ -110,22 +109,42 @@ case class PreAggregateTableHelper(
     // Datamap table name and columns are automatically added prefix with parent table name
     // in carbon. For convenient, users can type column names same as the ones in select statement
     // when config dmproperties, and here we update column names with prefix.
-    val longStringColumn = tableProperties.get(CarbonCommonConstants.LONG_STRING_COLUMNS)
+    // If longStringColumn is not present in dmproperties then we take long_string_columns from
+    // the parent table.
+    var longStringColumn = tableProperties.get(CarbonCommonConstants.LONG_STRING_COLUMNS)
+    val longStringColumnInParents = parentTable.getTableInfo.getFactTable.getTableProperties.asScala
+      .getOrElse(CarbonCommonConstants.LONG_STRING_COLUMNS, "").split(",").map(_.trim)
+    var varcharDatamapFields = ""
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Done!

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala (hunk 4 above)
Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Done!

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 8]
```diff
@@ -110,7 +110,29 @@ case class PreAggregateTableHelper(
     // Datamap table name and columns are automatically added prefix with parent table name
     // in carbon. For convenient, users can type column names same as the ones in select statement
     // when config dmproperties, and here we update column names with prefix.
-    val longStringColumn = tableProperties.get(CarbonCommonConstants.LONG_STRING_COLUMNS)
```
Comment (xuchuanyin) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: emm, actually this line of code was added to solve the same problem mentioned in the PR description. Previously we thought it's user's responsibility to explicitly specify the long_string_columns in the preagg datamap.

Please @kevinjmh also check this.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala [hunk 9]
```diff
@@ -110,7 +110,29 @@ case class PreAggregateTableHelper(
...
+    // If longStringColumn is not present in dm properties then we take long_string_columns from
+    // the parent table.
+    var longStringColumn = tableProperties.get(CarbonCommonConstants.LONG_STRING_COLUMNS)
```
Comment (kevinjmh) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: So user do not have to config dmproperties for long string anymore and this line is no use after this PR?

Comment (kevinjmh) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: The codes is not consistent with comments in line 113. Please fix it

Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Fixed it. If the user doesn't configure long_string in dmproperties, we take long_string_cols from the parent table.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala (hunk 8 above)
Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: This PR is for the scenario when user doesn't config the long_string_columns in dmprop then we take long_string_cols from the parent table.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2175/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10431/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2390/

**Comment by shardul-cr7** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2176/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2391/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10432/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala (hunk 9 above)
Comment (xuchuanyin) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: fine

**Comment by xuchuanyin** [comment]
LGTM

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateTableHelper.scala]: Done!

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2189/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10445/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2405/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2194/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2410/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10450/
